### PR TITLE
Recover and reconnect on restart

### PR DESCRIPTION
**Pull request by tcrain**
The previous pull request (https://github.com/SyncFree/antidote/pull/200) ensured that when an Antidote node failed and restarted it would load updates stored in the log into memory so that they can be read. 

This pull request continues on the recovery theme.  What is added here is that on restart, all previous connections and meta-data states are reestablished.  This includes the interDC connections, and the meta-data that is used to track the operation numbers so that the failed node can both send and receive correctly the messages that might have been lost during the failure.  This data is both loaded from the operation log and a dets file that keeps information about connections.
This pull request is resolving some of the issues described here https://github.com/SyncFree/antidote/issues/193

The recovery can be disabled by setting recover_meta_data_on_start to false in antidote.app.src https://github.com/SyncFree/antidote/blob/stable-meta-data/src/antidote.app.src#L15

New tests are added to test the basic functionality of the recovery.

### CODE REVIEW

File: riak_test/common.erl [hunk 1]
```diff
@@ -21,30 +21,42 @@
...
     end.
 
+just_clean_clusters(Clusters) ->
+    disconnect_dcs(Clusters),
+    lists:foldl(fun(Cluster, Acc) ->
+			rt:clean_cluster(Cluster),
+			Acc ++ [length(Cluster)]
+		end, [], Clusters),
+    ok.
+
```
Comment (bieniusa) [inline riak_test/common.erl]: The lists:foldl should be simplified to:

lists:map(fun(Cluster) -> rt:clean_cluster(Cluster))

You never use the accumulated value.

Comment (tcrain) [inline riak_test/common.erl]: Ok, this is fixed.

File: riak_test/common.erl [hunk 2]
```diff
@@ -21,30 +21,42 @@
...
   clean_clusters/1,
+  clean_clusters/2,
+  just_clean_clusters/1,
   setup_dc_manager/2]).
 
 %% It cleans the cluster and formes a new one with the same characteristic of the input one.
 %% In case the clean_cluster parameter is set to false in the riak_test configuration file
 %% the function simply returns the inputed set of clusters
 %% Clusters: A list of clusters as defined by rt
-clean_clusters(Clusters)->
```
Comment (bieniusa) [inline riak_test/common.erl]: I would suggest some renaming: 
clean_and_rebuild_clusters?

File: riak_test/multiple_dcs_node_failure_test.erl [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+    %% Kill and restart a node an be sure everyhing works
+    lager:info("Killing and restarting node ~w", [Node1]),
+    timer:sleep(5000),
+    rt:brutal_kill(Node1),
+    rt:start_and_wait(Node1),
+    rt:wait_until_ring_converged(Cluster1),
+    lager:info("Waiting until vnodes are restarted"),
+    rt:wait_until(Node1, fun wait_init:check_ready/1),
+    timer:sleep(10000),
```
Comment (bieniusa) [inline riak_test/multiple_dcs_node_failure_test.erl]: Are these sleeps really necessary? They are rather long.

File: src/wait_init.erl [hunk 4]
```diff
@@ -30,6 +30,7 @@
 %% and gen_servers serving read have been started.
 %% Returns true if all vnodes are initialized for all phyisical nodes,
 %% false otherwise
+-spec wait_ready_nodes([node()]) -> true.
 wait_ready_nodes([]) ->
```
Comment (bieniusa) [inline src/wait_init.erl]: Rewrite using lists:all  ?

File: src/wait_init.erl [hunk 5]
```diff
@@ -42,6 +43,7 @@ wait_ready_nodes([Node|Rest]) ->
 
 %% @doc This calls the check_ready function repatabliy
 %% until it returns true
+-spec wait_ready(node()) -> true.
 wait_ready(Node) ->
```
Comment (bieniusa) [inline src/wait_init.erl]: Rewrite using lists:all() ?

File: riak_test/multiple_dcs_node_failure_test.erl (hunk 3 above)
Comment (tcrain) [inline riak_test/multiple_dcs_node_failure_test.erl]: It's to be sure the updates have been stored to log durably.
Right now the disk log sync option is hardcoded in antidote.hrl, so maybe I should put that as one of the config options here: https://github.com/SyncFree/antidote/blob/master/src/antidote.app.src#L15
so then I can have the sync log be enabled for this test?

**Pull request merged**